File: src/main/java/action/fuse.java
package action;

import jason.asSemantics.DefaultInternalAction;
import jason.asSemantics.TransitionSystem;
import jason.asSemantics.Unifier;
import jason.asSyntax.*;

public class fuse extends DefaultInternalAction {

	private static final String ITEM_POS_MISSING = "The items has not been found in the positions list";

	@Override
	public Object execute( TransitionSystem ts, Unifier un, Term[] args) {

		/*List<String> items = split( args[ 0 ].toString( ) );
		List<String> positions = split( args[ 1 ].toString( ) );

		ListTerm result = new ListTermImpl( );
		result.addAll( items.stream( ).map( i -> buildLiteral( "item", new SimpleStructure[] {
				new SimpleStructure( "id", getValue( i, "id" ) ),
				new SimpleStructure( "quantity", Integer.parseInt( getValue( i, "quantity" ) ) ),
		}, split( splitStructAndList( positions.stream( )
				.filter( p -> Objects.equals( getValue( i, "id" ), getValue( p, "id" ) ) )
				.findFirst( ).orElseThrow( ( ) -> new NoSuchElementException( ITEM_POS_MISSING ) ) ).getValue( ) ).stream( )
				.map( p -> buildLiteral( "position", new SimpleStructure( "rack", getValue( p, "rack" ) ),
						new SimpleStructure( "shelf", getValue( p, "shelf" ) ),
						new SimpleStructure( "quantity", getValue( p, "quantity" ) )
				) ).collect( Collectors.toList( ) ).toArray( new Literal[] {} ) ) )
				.collect( Collectors.toList( ) ) );
		un.unifies( result, args[ 2 ] );
*/
		return true;
	}

}

File: src/main/kotlin/asl/action/fuse.kt
package asl.action

import jason.asSemantics.DefaultInternalAction
import jason.asSemantics.TransitionSystem
import jason.asSemantics.Unifier
import jason.asSyntax.ListTerm
import jason.asSyntax.ListTermImpl
import jason.asSyntax.Term
import common.translation.LiteralBuilder
import common.translation.LiteralBuilder.Companion.pairTerm
import common.translation.LiteralParser.getValue
import common.translation.LiteralParser.split
import common.translation.LiteralParser.splitStructAndList

class fuse : DefaultInternalAction() {
    override fun execute(ts: TransitionSystem, un: Unifier, args: Array<Term>): Any {
        val items: List<String> = split(args[0].toString())
        val positions: List<String> = split(args[1].toString())
        val result: ListTerm = ListTermImpl()
        result.addAll(items.map { i ->
            LiteralBuilder("item").setValues(
                    pairTerm("id", getValue(i, "id")!!),
                    pairTerm("quantity", getValue(i, "quantity")!!.toDouble())
            ).setQueue(*split(splitStructAndList(positions
                    .first { p: String? -> getValue(i, "id") == getValue(p!!, "id") }).second)
                    .map{ p -> LiteralBuilder("position").setValues(
                        pairTerm("rack", getValue(p, "rack")!!),
                        pairTerm("shelf", getValue(p, "shelf")!!),
                        pairTerm("quantity", getValue(p, "quantity")!!)
                    ).build() }.toTypedArray()
            ).build()
        } )
        return un.unifies(result, args[2])
    }

    companion object {
        private const val ITEM_POS_MISSING = "The items has not been found in the positions list"
    }
}